src/main/cljc/jarl/state.cljc

```diff
@@ -19,10 +19,8 @@
           (recur (next stack) aggregate))))))
 
 (defn get-local
-  ([state index]
-   (let [stack (get state :with-stack)
-         local (get state :local)
-         value (get local index)]
+  ([{stack :with-stack local :local} index]
```

Comment:
I'm not sure I'm on board with extracting map parameters directly in the argument list like this. Sure, all you need to know is that you need to pass an object that contains the correct stuff as the first arg, but you also lose the fact that all you need to pass is a `state`, and don't need to know the details of the state's internals to know that is what this function accepts.

Am I arguing for typing the function arg lists? 🤔

Reply:
At the very least, we should document these functions if we're not typing the args, so that others (that would include our future selves) can figure out when these functions are applicable.

Reply:
I'm fine with reverting the destructuring, although I rather like the explicitness here. We could definitely look into using something like [records](https://clojure.org/reference/datatypes) for things like the state here. Looks like they'd provide a good middle ground for stricter schema, but can still be references as maps (destructuring, etc).

Reply:
Lets explore this further. The discoverability of these functions suffers when all they take is a bag of random stuff you might know nothing about.

Reply:
👍 

Yeah, it's an interesting problem for sure! And I definitely get your point. We haven't really considered doc strings much before, but I wonder if that would be the sweet spot here? I.e. documenting the expected inputs in their entirety (e.g. `state`, etc), while having destructuring "document" what's actually _used_ from those.. I kinda like that a quick glance at function argument list will tell you exactly what is required from potentially large, nested objects, but I agree that knowing that these values in fact embed _state_ is useful as well.